test/llvmpasses/pipeline-o2-broadcast.jl

```diff
@@ -0,0 +1,109 @@
+# RUN: julia --startup-file=no -O2 --check-bounds=auto %s %t -O && llvm-link -S %t/* | FileCheck %s
+# RUN: julia --startup-file=no -O3 --check-bounds=auto %s %t -O && llvm-link -S %t/* | FileCheck %s
+
+include(joinpath("..", "testhelpers", "llvmpasses.jl"))
+
+# COM: Check broadcasted outer product is vectorized
+
+# COM: Float32
+# CHECK: @japi1_outer_product
+# CHECK: load <[[VEC_FACTOR:[0-9]+]] x float>
+# CHECK: fmul <[[VEC_FACTOR]] x float>
+# CHECK: store <[[VEC_FACTOR]] x float>
+
+# COM: Float64
+# CHECK: @japi1_outer_product
+# CHECK: load <[[VEC_FACTOR:[0-9]+]] x double>
+# CHECK: fmul <[[VEC_FACTOR]] x double>
+# CHECK: store <[[VEC_FACTOR]] x double>
+
+# COM: Int32
+# CHECK: @japi1_outer_product
+# CHECK: load <[[VEC_FACTOR:[0-9]+]] x i32>
+# CHECK: mul <[[VEC_FACTOR]] x i32>
+# CHECK: store <[[VEC_FACTOR]] x i32>
+
+# COM: Int64
+# CHECK: @japi1_outer_product
+# CHECK: load <[[VEC_FACTOR:[0-9]+]] x i64>
+# CHECK: mul <[[VEC_FACTOR]] x i64>
+# CHECK: store <[[VEC_FACTOR]] x i64>
+
+function outer_product(R, x, y)
+    R .= x .* y'
+end
+
+# COM: Check broadcasted inner product is vectorized
+
+# COM: Float32
+# CHECK: @japi1_inner_product
+# CHECK: load <[[VEC_FACTOR:[0-9]+]] x float>
+# CHECK: fmul <[[VEC_FACTOR]] x float>
+# CHECK: store <[[VEC_FACTOR]] x float>
+
+# COM: Float64
+# CHECK: @japi1_inner_product
+# CHECK: load <[[VEC_FACTOR:[0-9]+]] x double>
+# CHECK: fmul <[[VEC_FACTOR]] x double>
+# CHECK: store <[[VEC_FACTOR]] x double>
+
+# COM: Int32
+# CHECK: @japi1_inner_product
+# CHECK: load <[[VEC_FACTOR:[0-9]+]] x i32>
+# CHECK: mul <[[VEC_FACTOR]] x i32>
+# CHECK: store <[[VEC_FACTOR]] x i32>
+
+# COM: Int64
+# CHECK: @japi1_inner_product
+# CHECK: load <[[VEC_FACTOR:[0-9]+]] x i64>
+# CHECK: mul <[[VEC_FACTOR]] x i64>
+# CHECK: store <[[VEC_FACTOR]] x i64>
+
+function inner_product(R, x, y)
+    R .= x' .* y
```

Comment:
Personally, I'd call this an outer product.

I'd also call `squash` an elementwise product between a matrix and the outer product of two vectors.